# 1장 2강: 유의수준·검정력·표본 크기의 트레이드오프와 실질적 유의성 — 실습문제

## 실습 목표

- 효과 크기와 유의수준이 같을 때 표본 크기에 따라 검정력이 어떻게 변하는지 설명할 수 있다.
- 목표 검정력과 유의수준을 이용해 필요한 표본 크기를 계산할 수 있다.
- 두 집단의 평균 차이에 대한 Cohen’s d를 계산하고 해석할 수 있다.
- p-value와 효과 크기, 실무 기준을 함께 사용하여 의사결정 근거를 작성할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- statsmodels
- `ames_housing.csv`

주요 컬럼은 다음과 같습니다.

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `CentralAir` | 중앙 냉방시설 유무(`Y`, `N`) |
| `KitchenQual` | 주방 품질(`Ex`, `Gd`, `TA`, `Fa`) |

> 모든 검정은 별도 지시가 없으면 유의수준 `α = 0.05`를 사용합니다.  
> Cohen’s d는 절댓값을 기준으로 약 0.2는 작은 효과, 0.5는 중간 효과, 0.8 이상은 큰 효과로 해석합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.
from pathlib import Path
import pandas as pd
from scipy import stats

df = pd.read_csv("ames_housing.csv")

---

## 필수 1. 표본 크기와 검정력 비교

### 문제 1-1. 중앙 냉방시설 비교 연구에 필요한 표본 수 설계

#### 문제 설명

중앙 냉방시설이 있는 주택과 없는 주택의 판매가격을 비교하는 연구를 준비하고 있습니다. 사전 조사에서 표준화 효과 크기는 `0.3` 정도로 예상했습니다.

다음 조건에서 그룹당 표본 크기에 따른 검정력을 비교하고, 목표 검정력 0.8을 확보하기 위해 필요한 표본 수를 계산하세요.

- 예상 효과 크기: `0.3`
- 유의수준: `0.05`
- 양측검정
- 비교할 그룹당 표본 수: `30명`, `100명`, `300명`

#### 요구사항

1. `TTestIndPower()` 객체를 생성하세요.
2. 그룹당 표본 수가 30명, 100명, 300명일 때의 검정력을 각각 계산하세요.
3. 표본 크기별 검정력을 소수점 셋째 자리까지 출력하세요.
4. 유의수준 0.05, 목표 검정력 0.8, 효과 크기 0.3일 때 그룹당 필요한 표본 수를 계산하세요.
5. 계산된 표본 수는 소수점 이하를 올림하여 정수로 출력하세요.
6. 세 표본 크기 중 목표 검정력 0.8을 충족하는 경우를 확인하세요.


#### 해석 질문

**Q1.** 효과 크기와 유의수준이 같을 때 표본 크기가 커지면 검정력은 어떻게 변하나요?  
**Q2.** 검정력이 낮으면 실제 차이가 존재할 때 어떤 오류의 위험이 커지나요?  
**Q3.** 필요한 표본 수를 소수점 이하 올림으로 처리하는 이유는 무엇인가요?  
**Q4.** 표본 수를 무조건 크게 설정하는 것이 항상 최선인가요?

#### 제출 결과

- 표본 크기별 검정력
- 목표 검정력에 필요한 그룹당 표본 수
- 목표 검정력 충족 여부
- Q1~Q4 답변

In [ ]:
# 필수 1 코드를 작성하세요.

import math
from statsmodels.stats.power import TTestIndPower

# 1. 객체 생성
analysis = TTestIndPower()

effect_size = 0.3
alpha = 0.05

# 2~3. 표본 크기별 검정력
for n in [30, 100, 300]:
    power = analysis.power(effect_size=effect_size, nobs1=n, alpha=alpha)
    print(f"n={n}: 검정력={power:.3f}")

# 4~5. 필요 표본 수
n_needed = analysis.solve_power(effect_size=effect_size, power=0.8, alpha=alpha)
print("필요 표본 수:", math.ceil(n_needed))

# 6. 목표 검정력 충족 여부
for n in [30, 100, 300]:
    power = analysis.power(effect_size=effect_size, nobs1=n, alpha=alpha)
    print(f"n={n}:", "충족" if power >= 0.8 else "미충족")


n=30: 검정력=0.208
n=100: 검정력=0.560
n=300: 검정력=0.956
필요 표본 수: 176
n=30: 미충족
n=100: 미충족
n=300: 충족


### 필수 1 답변 작성란

- **Q1.**
-> 표본 크기가 커지면 일반적으로 검정력이 높아진다.
-> 표본이 많아질수록 실제 차이를 우연한 변동과 구분하기 쉬워짐.

- **Q2.**
-> 실제 차이가 있는데도 귀무가설을 기각하지 못하는 제 2종 오류의 위험이 커짐
-> 예를 들어 검정력이 0.8이라면 가정한 실제 차이를 놓칠 확률인 '베타'는 '1 - 0.8 = 0.2'입니다.

- **Q3.**
-> 사람이나 주택을 소수 단위로 모집할 수 없고
-> 내림하면 계산된 목표 검정력을 충족하지 못할 수 있다.

- **Q4.**
-> 아닙니다. 표본 확보엥는 비용과 시간이 소요가 됨


---

## 필수 2. 통계적 유의성과 실질적 유의성 종합 판단

### 문제 2-1. 중앙 냉방시설 유무에 따른 판매가격 차이 분석

#### 문제 설명

한 부동산 회사는 중앙 냉방시설이 있는 주택과 없는 주택의 평균 판매가격 차이를 분석하려고 합니다. 회사는 두 집단의 평균 판매가격 차이가 **100,000달러 이상**이어야 냉방시설 설치 지원 사업을 검토할 실무적 가치가 있다고 정했습니다.

#### 요구사항

1. `CentralAir == "Y"`인 주택의 `SalePrice`를 `air_yes`에 저장하세요.
2. `CentralAir == "N"`인 주택의 `SalePrice`를 `air_no`에 저장하세요.
3. 두 집단의 표본 수와 평균 판매가격을 출력하세요.
4. `stats.ttest_ind()`로 두 집단 평균 차이에 대한 양측검정을 수행하세요.
5. 제공된 공식에 따라 Cohen’s d 계산 함수를 작성하고 효과 크기를 구하세요.
6. 평균 차이, p-value, Cohen’s d를 출력하세요.
7. 다음 기준으로 결과를 판단하세요.
   - `p-value ≤ 0.05`: 통계적으로 유의함
   - `|Cohen’s d| ≥ 0.8`: 큰 효과
   - `|평균 차이| ≥ 100000`: 회사 기준에서 실질적으로 유의함
8. 세 결과를 종합하여 사업 검토 여부에 대한 근거를 작성하세요.

#### Cohen’s d 공식

두 집단의 평균 차이를 합동 표준편차로 나누어 계산합니다.

$$
d = \frac{\bar{x}_1 - \bar{x}_2}{s_p}
$$

$$
s_p = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}}
$$

- $\bar{x}_1$, $\bar{x}_2$: 각 집단의 평균
- $s_1$, $s_2$: 각 집단의 표본 표준편차
- $n_1$, $n_2$: 각 집단의 표본 수
- $s_p$: 합동 표준편차

#### 해석 질문

**Q1.** p-value와 Cohen’s d는 각각 어떤 정보를 제공하나요?  
**Q2.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?  
**Q3.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  
**Q4.** 회사가 정한 100,000달러 기준에서 실질적으로 유의한가요?  
**Q5.** 통계적으로 유의하고 효과 크기가 크더라도 회사의 실무 기준을 충족하지 못할 수 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립표본 t검정 결과
- 평균 차이와 Cohen’s d
- 통계적·효과크기·회사 기준 판단
- 최종 의사결정 근거
- Q1~Q5 답변

In [8]:
# 필수 2 코드를 작성하세요.

import numpy as np
from scipy import stats

# 1~2. 집단 분리
air_yes = df.loc[df["CentralAir"] == "Y", "SalePrice"]
air_no  = df.loc[df["CentralAir"] == "N", "SalePrice"]

# 3. 표본 수와 평균
print(f"Y: n={len(air_yes)}, 평균={air_yes.mean():.2f}")
print(f"N: n={len(air_no)}, 평균={air_no.mean():.2f}")

# 4. 양측 독립표본 t-검정
t_stat, p_val = stats.ttest_ind(air_yes, air_no, alternative="two-sided")

# 5. Cohen's d 함수
def cohens_d(x, y):
    n1, n2 = len(x), len(y)
    s1, s2 = x.std(ddof=1), y.std(ddof=1)
    pooled_std = np.sqrt(((n1 - 1) * s1**2 + (n2 - 1) * s2**2) / (n1 + n2 - 2))
    return (x.mean() - y.mean()) / pooled_std

d = cohens_d(air_yes, air_no)

# 6. 출력
print(f"평균 차이: {air_yes.mean() - air_no.mean():.2f}")
print(f"t={t_stat:.4f}, p={p_val:.4f}")
print(f"p = {p_val:.3e}")
print(f"Cohen's d: {d:.4f}")

Y: n=1365, 평균=186186.71
N: n=95, 평균=105264.07
평균 차이: 80922.64
t=9.9149, p=0.0000
p = 1.810e-22
Cohen's d: 1.0521


In [7]:
result = stats.ttest_ind(air_yes, air_no)
print(result)

TtestResult(statistic=np.float64(9.914905121389344), pvalue=np.float64(1.8095061559266053e-22), df=np.float64(1458.0))


### 필수 2 답변 작성란

- **Q1.**
-> p-value는 관측된 차이가 통계적으로 유의한지 판단하는 데 사용
-> Cohen's d는 두 집단 차이의 크기를 표준화하여 보여줌

- **Q2.**
-> 유의함.p-value가 0.05보다 작음
-> 귀무가설 : 냉방 시설 유무는 주택 가격 평균에 차이를 보이지 않는다 -> 기각

- **Q3.**
-> Cohen's의 절대값이 1.05고 규정한 0.8보다 크기 때문에 큰 효과로 해석할 수 있다.

- **Q4.**
-> -> 유의하지 않음, 80.900달러 정도였기 때문에 회사의 기대 미충족

- **Q5.**
-> 그렇다. 통계적 유의성은 차이가 있다는 근거가 있는가?
-> 효과 크기는 자료의 흩어짐에 비해 차이가 큰가?
-> 회사 입장은 업무에서 요구하는 최소 금액 차이를 넘는가?


---

## 과제. 주방 품질에 따른 판매가격 차이 분석

### 문제 3-1. 주방 품질이 `Gd`인 집단과 `TA`인 집단 비교

#### 문제 설명

부동산 중개회사는 주방 품질이 `Gd`(Good)인 주택과 `TA`(Typical/Average)인 주택의 평균 판매가격을 비교하려고 합니다. 회사는 평균 판매가격 차이가 **50,000달러 이상**이면 마케팅에서 주방 품질의 차이를 강조할 실무적 가치가 있다고 판단합니다.

> 필수 2에서 학습한 검정과 효과 크기 계산 절차를 새로운 두 집단에 적용하는 과제입니다.

#### 요구사항

1. 주방 품질이 `Gd`인 집단과 `TA`인 집단의 `SalePrice`를 각각 준비하세요.
2. 두 집단의 표본 수와 평균 판매가격을 출력하세요.
3. 두 집단의 평균 차이에 대한 양측 독립표본 t검정을 수행하세요.
4. Cohen’s d를 계산하세요.
5. 평균 차이, p-value, Cohen’s d를 출력하세요.
6. 다음 기준에 따라 각각 판단하세요.
   - 통계적 유의성: `p-value ≤ 0.05`
   - 큰 효과: `|Cohen’s d| ≥ 0.8`
   - 실무적 유의성: `|평균 차이| ≥ 50000`
7. 세 가지 판단을 종합하여 주방 품질을 마케팅에서 강조할 근거가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?  
**Q2.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  
**Q3.** 회사가 정한 50,000달러 기준을 충족하나요?  
**Q4.** 최종적으로 주방 품질 차이를 마케팅에서 강조할 근거가 있다고 볼 수 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- t통계량과 p-value
- 평균 차이와 Cohen’s d
- 통계적·실질적 유의성 판단
- 최종 결론
- Q1~Q4 답변

In [ ]:
# 과제 코드를 작성하세요.

import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv('ames_housing.csv')

# 1. 두 집단 준비
gd = df.loc[df['KitchenQual'] == 'Gd', 'SalePrice'].dropna()
ta = df.loc[df['KitchenQual'] == 'TA', 'SalePrice'].dropna()

# 2. 표본 수와 평균
n1, n2 = len(gd), len(ta)
m1, m2 = gd.mean(), ta.mean()
s1, s2 = gd.std(ddof=1), ta.std(ddof=1)
print(f"Gd 집단 : n = {n1}, 평균 = {m1:,.2f}")
print(f"TA 집단 : n = {n2}, 평균 = {m2:,.2f}")

# 3. 독립표본 t검정 (양측, Welch)
t_stat, p_val = stats.ttest_ind(gd, ta, equal_var=False)

# 4. Cohen's d (합동표준편차 사용)
sp = np.sqrt(((n1-1)*s1**2 + (n2-1)*s2**2) / (n1+n2-2))
diff = m1 - m2
d = diff / sp

# 5. 결과 출력
print(f"\n평균 차이  : {diff:,.2f}")
print(f"t 통계량   : {t_stat:.4f}")
print(f"p-value    : {p_val:.4g}")
print(f"Cohen's d  : {d:.4f}")

# 6. 세 가지 기준 판단
print(f"\n통계적 유의성 (p ≤ 0.05)  : {'충족' if p_val <= 0.05 else '미충족'}")
print(f"큰 효과 (|d| ≥ 0.8)        : {'충족' if abs(d) >= 0.8 else '미충족'}")
print(f"실무적 유의성 (차이 ≥ 5만) : {'충족' if abs(diff) >= 50000 else '미충족'}")


Gd 집단 : n = 586, 평균 = 212,116.02
TA 집단 : n = 735, 평균 = 139,962.51

평균 차이  : 72,153.51
t 통계량   : 23.9812
p-value    : 4.933e-99
Cohen's d  : 1.3991

통계적 유의성 (p ≤ 0.05)  : 충족
큰 효과 (|d| ≥ 0.8)        : 충족
실무적 유의성 (차이 ≥ 5만) : 충족


### 과제 답변 작성란

- **Q1.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?
-> 유의하다. 

- **Q2.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?
-> 큰 효과에 해당한다. Cohen의 관례상 0.2 = 작음, 0.5 = 중간, 0.8 = 큼이므로,  1.4는 큼 기준을 훨씬 넘어서는 수준이다.

- **Q3.** 회사가 정한 50,000달러 기준을 충족하나요?
-> 평균 판매 가격 차이가 50,000 달러를 초과하므로 충족한다.

- **Q4.** 최종적으로 주방 품질 차이를 마케팅에서 강조할 근거가 있다고 볼 수 있나요?
-> 통계적 유의성, 큰 효과 크기, 실무적 기준을 모두 충족하므로 근거는 있다. 

---

## 실습 마무리

1. 검정력은 무엇이며 제2종 오류와 어떤 관계가 있나요?
2. 같은 효과 크기와 유의수준에서 표본 크기가 커지면 검정력은 어떻게 변하나요?
3. 목표 검정력이 높거나 발견하려는 효과가 작을수록 필요한 표본 수는 어떻게 변하나요?
4. p-value와 Cohen’s d를 함께 확인해야 하는 이유는 무엇인가요?
5. 통계적으로 유의한 결과가 반드시 실무적으로 중요한 결과를 의미하나요?